### DISCUSSION TITLE
Prioritizing job

### DISCUSSION THREAD

**Discussion by stephenhuey** [bensheldon/good_job#524] [Q&A]
Good Job is working well for me!  One thing I'm trying to sort through is the sparse documentation on Active Job prioritization in the Rails 6 guide and a lot of old discussions online related to DJ and others.  

Can you give me concrete steps to have at least priority levels?  For example, sometimes we do a lot of data processing, and Active Storage might enqueue a bunch of Analyze jobs in the midst of this, but I want to ensure the core business data processing gets worked before the Analyze jobs.  Also, if a user comes along during this work and tries to reset their password, I want that email job worked ahead of anything else going on! I'm looking for code examples of how to handle this.  

So, I know AJ/GJ both have some notion of named queues and numerical priorities.  I believe AJ lets you define both queue and priority on a job class - does GJ respect both?  I imagine in a Heroku Procfile you can limit some workers to certain queues and other workers to other queues, but I'm not sure if I'd want to bother with that if I can just make sure priorities are honored.  Can I get Active Storage to assign a lower priority to Analyze jobs? 

In the GJ example app, I see a lot of jobs have negative priorities - negative is a higher priority than zero, right?  So if I make my important jobs a lower number than Analyze jobs, then they'll get executed first without me having to touch Active Storage jobs, right?   

Thanks for any help!

**Comment by bensheldon** [comment]
Great question! But a dense one. Let me try to tackle it:

- GoodJob respects both queues and priorities.
- GoodJob orders jobs within a queue (or pool of queues) by descending priority (e.g. a job with priority 10 will execute before priority -10). I think this is different than DJ in which a priority is more of a filter.
- The advice that I read in Nate Berkopec's ["Sidekiq in Practice"](https://nateberk.gumroad.com/l/sidekiqinpractice) is to use separate queues based on the allowed total latency (e.g. queuing latency plus service latency). e.g. in your example, you want the password reset emails to go out `:asap` whereas the analyze jobs are maybe `:ten_minutes`. 
- The danger of putting everything on the same queue is that if all the executors are busy with Analyze jobs, each of which takes 5 minutes to process, your password reset emails could take up to 5 minutes to be executed regardless of their relative priority.
- I personally run GoodJob with something like `--queues=mice:2;elephants:2`. Mice are my asap jobs (fast, brief, immediate), and elephants (slow, long, bulked) are my whenever jobs. I'll sometimes set a priority within `:elephants`, but not mice because I don't ever want that queue's pool to be saturated.

**Reply by stephenhuey** [reply]
Thank you so much!  I meant to reply right away.  

Since I'm not sure I can easily control the queue or priority for something like Active Storage Analyze jobs, and I don't ever want those to get in the way of my mice/asap jobs, is the easiest way to keep Analyze jobs or big/slow jobs from interfering to just assign a high number like 10 to my mice/asap jobs?  If I understood your response, do you mean that GJ will next work the highest priority job out of a pool of queues?  

Also, I'm trying to figure out what the thread count does.  In your preferred example, does that mean each worker has 2 threads that work the mice queue and 2 that work the elephants queue?  If so, do those threads work ONLY those queues?  

Much obliged,
Stephen

**Reply by bensheldon** [reply]
You can change the queue for [ActiveStorage jobs via configuration](https://guides.rubyonrails.org/configuring.html#config-active-storage-queues-analysis):

```ruby
config.active_storage.queues.analysis = :elephants
config.active_storage.queues.purge = :elephants
config.action_mailbox.queues.incineration = :elephants
config.action_mailbox.queues.routing = :elephants
```

You could also set an initializer to change the priorities directly on [ActiveStorage Jobs](https://github.com/rails/rails/tree/f01fee7fc49371c28a1a59ed406f7ee00fcda3bf/activestorage/app/jobs/active_storage), e.g. `ActiveStorage::AnalyzeJob.queue_with_priority(50)`.

Recommendation: Use queues instead of priorities.

In my example `--queues=mice:2;elephants:2`, that would mean that there would be 2 threads that only work mice, and 2 threads that only worked elephants. 

Thinking more about it now, I recommend something like this: `--queues=elephants:2;-elephants:2`. Then you'd have 2 threads for the elephants queue, and you'd have 2 threads for jobs on all other queues ("minus elephants"). Put the ActiveStorage jobs on the `:elephants` queue (or `:low_priority` or `:eventually`).

**Comment by segiddins** [comment]
> negative is a higher priority than zero, right?

I think that's how it's documented to work for active job, but the query for executions (`. priority_ordered`) is `Order executions by priority (highest priority first).`, so im confused how it's supposed to work in practice

**Reply by bensheldon** [reply]
Good question. I think Active Job leaves it undefined/up to the adapter, but I'd love a link if you can find it documented.

**Good Job treats it as "higher number is higher priority".**

...I also previously worked in Drupal which used "weight" where "lower weights float to the top" (which is the opposite of "priority")

From some quick research, TIL that GoodJob does the opposite of other adapters. Both Que and Delayed both use lower-is-sooner. Que at least explains "We use the Linux priority scale - a lower number is more important."

**Reply by segiddins** [reply]
https://github.com/rails/rails/blob/e17faead4f2aff28da079d50f02ea5b015322d5b/activejob/lib/active_job/core.rb#L22

**Reply by bensheldon** [reply]
oh that's interesting. Thank you! This will be a fun deprecation cycle 😅 

Would you make an issue for that?

**Reply by bensheldon** [reply]
I made a PR: https://github.com/bensheldon/good_job/pull/883. Would love any feedback, especially on the "smaller number is higher priority" naming.

**Comment by ScotterC** [comment]
Less related but seems the best place to ask. When does `queue_name_prefix` get computed in relation to the `queues` configuration. If I have `config.active_job.queue_name_prefix = Rails.env` and config `--queues=elephants:2;-elephants:2` does that mean `production_elephants` will have two threads on it? Or do I need to take the prefix into account when configuring queues? I couldn't easily find this in the docs so I'm going to take a look at the code.

Update:

[queue_name_prefix](https://edgeguides.rubyonrails.org/active_job_basics.html#the-purpose-of-active-job) is an ActiveJob config so  it'll apply after GoodJob config.

**Reply by bensheldon** [reply]
TIL about `queue_name_prefix`. GoodJob doesn't take that into account so you'd need to manually include it in your `--queues=` configuration.

Why are you using that configuration option btw?

**Reply by ScotterC** [reply]
I got one clear benefit from it. Working on an app small enough that I can dump the production db and load it locally right now. Prefixing Rails.env for each queue means the production jobs don't startup in my local environment after dumping and loading that db.
